enterprise/coderd/prebuilds/metricscollector_test.go

```diff
@@ -1,73 +1,247 @@
+								}
+							}
+
+							collector := prebuilds.NewMetricsCollector(db, logger, reconciler)
+							registry := prometheus.NewPedanticRegistry()
+							registry.Register(collector)
+
+							numTemplates := 2
+							for i := 0; i < numTemplates; i++ {
+								orgID, templateID := setupTestDBTemplate(t, db, ownerID)
+								templateVersionID := setupTestDBTemplateVersion(t, ctx, db, pubsub, orgID, ownerID, templateID)
+								preset := setupTestDBPreset(t, ctx, db, pubsub, templateVersionID, 1, uuid.New().String())
+								setupTestDBPrebuild(
+									t, ctx, db, pubsub,
+									transition, jobStatus, orgID, templateID, templateVersionID, preset.ID, initiatorID, ownerID,
+								)
+							}
+
+							metricsFamilies, err := registry.Gather()
+							require.NoError(t, err)
+
+							templates, err := db.GetTemplates(ctx)
+							require.NoError(t, err)
+							require.Equal(t, numTemplates, len(templates))
+
+							for _, template := range templates {
+								templateVersions, err := db.GetTemplateVersionsByTemplateID(ctx, database.GetTemplateVersionsByTemplateIDParams{
+									TemplateID: template.ID,
+								})
+								require.NoError(t, err)
+								require.Equal(t, 1, len(templateVersions))
+
+								presets, err := db.GetPresetsByTemplateVersionID(ctx, templateVersions[0].ID)
+								require.NoError(t, err)
+								require.Equal(t, 1, len(presets))
+
+								for _, preset := range presets {
+									checks := []metricCheck{
+										{"coderd_prebuilds_created", test.shouldIncrementPrebuildsCreated, true},
+										{"coderd_prebuilds_failed", test.shouldIncrementPrebuildsFailed, true},
+										{"coderd_prebuilds_assigned", test.shouldIncrementPrebuildsAssigned, true},
+										{"coderd_prebuilds_desired", test.shouldSetDesiredPrebuilds, false},
+										{"coderd_prebuilds_actual", test.shouldSetActualPrebuilds, false},
+										{"coderd_prebuilds_eligible", test.shouldSetEligiblePrebuilds, false},
+									}
+
+									labels := map[string]string{
+										"template_name": template.Name,
+										"preset_name":   preset.Name,
+									}
+
+									for _, check := range checks {
+										metric := findMetric(metricsFamilies, check.name, labels)
+										if check.value == nil {
+											continue
+										}
+
+										require.NotNil(t, metric, "metric %s should exist", check.name)
+										if check.isCounter {
+											require.Equal(t, *check.value, metric.GetCounter().GetValue(), "counter %s value mismatch", check.name)
+										} else {
+											require.Equal(t, *check.value, metric.GetGauge().GetValue(), "gauge %s value mismatch", check.name)
+										}
+									}
+								}
+							}
+						})
+					}
+				}
+			}
+		}
+	}
+}
+
+func findMetric(metricsFamilies []*prometheus_client.MetricFamily, name string, labels map[string]string) *prometheus_client.Metric {
+	for _, metricFamily := range metricsFamilies {
+		if metricFamily.GetName() == name {
```

Comment:
Nit: keep happy path unintended; i.e. continue if the name doesn't match

Reply:
+1